p2p/src/pubsub.rs

```diff
@@ -149,90 +111,142 @@ where
             .await
             .map_err(|_| P2pError::SubsystemFailure)?;
 
+        self.pubsub_handle.subscribe(&self.topics).await?;
+        Ok(rx)
+    }
+
+    /// Process block announcement from the network
+    async fn process_block_announcement(
+        &mut self,
+        peer_id: T::PeerId,
+        message_id: T::MessageId,
+        block: Block,
+    ) -> crate::Result<()> {
+        let result = self
+            .chainstate_handle
+            .call_mut(move |this| this.process_block(block, chainstate::BlockSource::Peer))
+            .await?;
+
+        let (validation_result, score) = match result {
+            Ok(_) => (ValidationResult::Accept, 0),
+            Err(ProcessBlockError(ref block_error)) => match block_error {
```

Comment:
I'm confused. Why don't you directly call ban_score on the BlockError object? Why match all these error cases? You shouldn't care. Just get the ban score and add it to the total. If it's higher than the max, ban the dude. The max should come from command line arguments, btw. That's still not implemented. Currently 100 is fine for default.

Reply:
I need to match `BlockAlreadyExists` either way but I think `ValidationResult` here is more interesting than the ban score. The easiest way to handle is that if we have an error and ban score is 0, the block is ignored, if it's more than 0, the block is rejected and peer is banned. I'm not opposed to switching to that. But some of these errors might be totally our fault in which case we shouldn't ignore the block. I don't know

Reply:
Let's revisit this issue later when the implementation is complete to see how best to tackle this. At best, I'd suggest separating the score calculation from the match. It seems too convoluted while you can simply just call a method on the error to get that number.